master/buildbot/test/util/db.py

```diff
@@ -65,16 +65,76 @@ class RealDatabaseMixin(object):
+        #
+        # Dropping schema from model is correct and working operation only
+        # if database schema is exactly corresponds to the model schema.
+        #
+        # If it is not (e.g. migration script failed or migration results in
+        # old version of model), then some tables outside model schema may be
+        # present, which may reference tables in the model schema.
+        # In this case either dropping model schema will fail (if database
+        # enforces referential integrity, e.g. PostgreSQL), or
+        # dropping left tables in the code below will fail (if database allows
+        # removing of tables on which other tables have references,
+        # e.g. SQLite).
+        #
+        # 2. Introspect database contents and drop found tables.
+        #
+        #     meta = MetaData(bind=conn)
+        #     meta.reflect()
+        #     meta.drop_all()
+        #
+        # May fail if schema contains reference cycles (and Buildbot schema
+        # has them). Reflection looses metadata about how reference cycles
+        # can be teared up (e.g. use_alter=True).
+        # Introspection may fail if schema has invalid references
+        # (e.g. possible in SQLite).
+        #
+        # 3. What is actually needed here is accurate code for each engine
+        # and each engine configuration that will drop all tables,
+        # indexes, constraints, etc in proper order or in a proper way
+        # (using tables alternation, or DROP TABLE ... CASCADE, etc).
+        #
+        # Conclusion: use approach 2 with manually teared apart known
+        # reference cycles.
 
-        # see if we can find any other tables to drop
         try:
             meta = MetaData(bind=conn)
+
+            # Reflect database contents. May fail, e.g. if table references
+            # non-existent table in SQLite.
             meta.reflect()
+
+            # Table.foreign_key_constraints introduced in SQLAlchemy 1.0.
+            if sa_version()[:2] >= (1, 0):
+                # Restore `use_alter` settings to break known reference cycles.
+                # Main goal of this part is to remove SQLAlchemy warning
+                # about reference cycle.
+                # Looks like it's OK to do it only with SQLAlchemy >= 1.0.0,
+                # since it's not issued in SQLAlchemy == 0.8.0
+                if 'buildsets' in meta.tables:
+                    buildsets_table = meta.tables['buildsets']
+                    for fkc in buildsets_table.foreign_key_constraints:
+                        if fkc.referred_table.name == 'builds':
+                            fkc.use_alter = True
+
+                if 'builds' in meta.tables:
+                    builds_table = meta.tables['builds']
+                    for fkc in builds_table.foreign_key_constraints:
+                        if fkc.referred_table.name == 'buildrequests':
+                            fkc.use_alter = True
```

Comment:
I'd rather have this read from some data structure like `[('buildsets', 'builds'), ('builds', 'buildrequests')]`